src/Controller/ContentTypeController.php

```diff
@@ -0,0 +1,119 @@
+     */
+    public const string PATH_PATTERN = '([\p{L}\p{N}]+(?:-[\p{L}\p{N}]+)*(?:\/[\p{L}\p{N}]+(?:-[\p{L}\p{N}]+)*)*\/?)$';
+    private Version $version;
+
+    public function __construct(
+        private ContentTypeControllerRegistry $registry,
+        private ContainerInterface $container,
+        private LoggerInterface $logger,
+        private RequestHandlerInterface $requestHandler,
+        private bool $public,
+        private bool $mustRevalidate,
+        private int $maxAge,
+        private int $smaxAge,
+        string $version,
+    ) {
+        $this->version = Version::from($version);
+    }
+
+    public function __invoke(Request $request, string $slug): Response
+    {
+        $response = $this->requestHandler->handle($request, $slug);
+
+        if ($response instanceof Response) {
+            return $response;
+        }
+
+        $story = $response->story;
+        Assert::keyExists($story, 'published_at');
+        Assert::keyExists($story, 'content');
+        Assert::keyExists($story['content'], 'component');
+
+        try {
+            $definition = $this->registry->bySlug(\sprintf('/%s', $slug));
+            $this->logger->debug('Found content type controller by slug.', [
+                'slug' => $slug,
+                'controller' => $definition->className,
+            ]);
+        } catch (ContentTypeControllerNotFoundException) {
+            $type = TrimmedNonEmptyString::fromString($story['content']['component'])->toString();
+            $definition = $this->registry->byType($type);
+            $this->logger->debug('Found content type controller by component.', [
+                'type' => $type,
+                'controller' => $definition->className,
+            ]);
+        }
+
+        /** @var callable(Request, ContentTypeInterface): Response $controller */
+        $controller = $this->container->get($definition->className);
+
+        /** @var ContentTypeInterface $contentType */
+        $contentType = new $definition->dto($story);
+
+        $this->logger->debug('Calling content type controller.', [
+            'controller' => $definition->className,
+            'content_type' => $contentType,
+        ]);
+
+        $response = $controller($request, $contentType);
+
+        $this->logger->debug('Content type controller returned response.', [
+            'controller' => $definition->className,
+            'response' => $response,
+        ]);
+
+        if ($this->version->equals(Version::Published)) {
+            if ($this->public) {
+                $response->setPublic();
+            } else {
+                $response->setPrivate();
+            }
+
+            if ($this->mustRevalidate) {
+                $response->mustRevalidate();
+            }
+
+            $response->setMaxAge($this->maxAge);
+            $response->setSharedMaxAge($this->smaxAge);
+            $response->setLastModified(new DateTimeImmutable($story['published_at']));
```

Comment:
I think all this should be done only if not already set in response, allowing custom contenttype controller to define dynamically those values (similar to what's done in symfony/something ^^)

Reply:
https://github.com/symfony/symfony/blob/a2a47bc07b9997d663409fe0b97fa753a656036b/src/Symfony/Component/HttpKernel/EventListener/CacheAttributeListener.php#L94